core/trino-main/src/main/java/io/trino/operator/output/PositionsAppenderPageBuilder.java

```diff
@@ -98,7 +99,7 @@ private void declarePositions(int positions)
 
     public boolean isFull()
     {
-        return declaredPositions == Integer.MAX_VALUE || getSizeInBytes() >= maxPageSizeInBytes;
+        return declaredPositions >= PageProcessor.MAX_BATCH_SIZE || getSizeInBytes() >= maxPageSizeInBytes;
```

Comment:
this was mimicking `io.trino.spi.PageBuilder#isFull`. I think `PageProcessor.MAX_BATCH_SIZE` might be ok for internal processing but could be too small for remote pages. I would set it to something bit more conservative, like `64_000` or `32_000`

Reply:
> this was mimicking `io.trino.spi.PageBuilder#isFull`
Agreed, that's definitely where this came from, but there were two problems with doing that:
- It failed to account for the fact that `PageBuilder` is meant to be used in a loop appending one row at a time checking `isFull()` on each iteration
- `PageBuilder` doesn't work over dictionary or RLE blocks, so the assumption is that the maximum size in bytes will be reached at some point much earlier than `Integer.MAX_VALUE`

> I think `PageProcessor.MAX_BATCH_SIZE` might be ok for internal processing but could be too small for remote pages. I would set it to something bit more conservative, like `64_000` or `32_000`

I'm not sure I understand why we should want to make a distinction between the two. Ultimately, remote pages turn into local pages on the receiving end and `PageProcessor.MAX_BATCH_SIZE` is chosen to attempt to balance the per-page overheads against per-page processing time in downstream operators. The transport sends multiple pages per response, so I don't think there's an inherent benefit to sending oversized pages over the wire compared to multiple smaller ones (as long as they're not too small).

Reply:
Per our conversation offline, partitioned output pages will go through a fragmenting local exchange on the receiving side and so should avoid flushing as eagerly. I've increased the threshold to 4x `MAX_BATCH_SIZE` in the latest revision.